scalding-serialization/src/main/scala/com/twitter/scalding/serialization/Boxed.scala

```diff
@@ -801,10 +801,27 @@ object Boxed {
 
   def allClasses: Seq[Class[_ <: Boxed[_]]] = allBoxes.map(_._2)
 
-  def next[K]: (K => Boxed[K], Class[Boxed[K]]) = boxes.get match {
+  private[this] val boxedCache = new java.util.concurrent.ConcurrentHashMap[AnyRef, (Any => Boxed[Any], Class[Boxed[Any]])]()
+
+  def next[K](cacheKey: Option[AnyRef]): (K => Boxed[K], Class[Boxed[K]]) =
+    cacheKey match {
+      case Some(cls) =>
+        val untypedRes = Option(boxedCache.get(cls)) match {
+          case Some(r) => r
+          case None =>
+            val r = fetchNextBoxed[K]()
+            boxedCache.putIfAbsent(cls, r.asInstanceOf[(Any => Boxed[Any], Class[Boxed[Any]])])
+            r
+        }
+        untypedRes.asInstanceOf[(K => Boxed[K], Class[Boxed[K]])]
+      case None =>
+        fetchNextBoxed[K]()
+    }
+
+  private[scalding] def fetchNextBoxed[K](): (K => Boxed[K], Class[Boxed[K]]) = boxes.get match {
```

Comment:
can we keep this called `next` for binary compat and have a `defNextCached(o: EquivSerialization[_]): ...` which handles the cache logic?